### PR TITLE
MNIST tutorial

### PR DESCRIPTION
**Pull request by crangelsmith**
Adding a short tutorial about running Affinity VAE on MNIST dataset.

This tutorial contains the following:

- Example of how to create a rotated MNIST dataset for affinityVAE
- Example of how to train affinity on the MNIST rotated data.
- Brief description of the outputs

I have tested it myself and it runs without issues, but it would be great to have an external reviewer less familiar with affinity to make sure everything makes sense and each steps runs without bugs.

### CODE REVIEW

**Comment by parkyed** [comment]
Hi @crangelsmith , in working through the tutorial, in running the following line

python mnist_saver.py --mnist_file mnist.pkl.gz --output_path .

Traceback (most recent call last):
  File "/Users/ep/Documents/1_datasets/aff_vae/affinity-vae/tutorials/mnist_saver.py", line 179, in <module>
    mnist_saver.run()
  File "/Users/ep/Documents/1_datasets/aff_vae/affinity-vae/tutorials/mnist_saver.py", line 128, in run
    image = rotate(
  File "/Users/ep/Documents/1_datasets/aff_vae/env/lib/python3.9/site-packages/scipy/ndimage/_interpolation.py", line 894, in rotate
    raise ValueError('input array should be at least 2D')
ValueError: input array should be at least 2D

Environment details

absl-py==2.0.0
-e git+https://github.com/alan-turing-institute/affinity-vae.git@36ff6a1d20f77b9594d69485ec4b7814836e8ab1#egg=affinivae
altair==5.2.0
annotated-types==0.6.0
attrs==23.1.0
cachetools==5.3.2
certifi==2023.11.17
charset-normalizer==3.3.2
click==8.1.7
contourpy==1.2.0
cycler==0.12.1
filelock==3.13.1
fonttools==4.45.1
fsspec==2023.10.0
google-auth==2.23.4
google-auth-oauthlib==1.1.0
grpcio==1.59.3
idna==3.6
imageio==2.33.0
importlib-metadata==6.8.0
importlib-resources==6.1.1
Jinja2==3.1.2
joblib==1.3.2
jsonschema==4.20.0
jsonschema-specifications==2023.11.1
kiwisolver==1.4.5
lazy_loader==0.3
llvmlite==0.41.1
Markdown==3.5.1
MarkupSafe==2.1.3
matplotlib==3.8.2
mpmath==1.3.0
mrcfile==1.4.3
networkx==3.2.1
numba==0.58.1
numpy==1.26.2
oauthlib==3.2.2
packaging==23.2
pandas==2.1.3
Pillow==10.1.0
protobuf==4.23.4
pyasn1==0.5.1
pyasn1-modules==0.3.0
pydantic==2.5.2
pydantic_core==2.14.5
pynndescent==0.5.11
pyparsing==3.1.1
python-dateutil==2.8.2
pytz==2023.3.post1
PyYAML==6.0.1
referencing==0.31.0
requests==2.31.0
requests-oauthlib==1.3.1
rpds-py==0.13.1
rsa==4.9
scikit-image==0.22.0
scikit-learn==1.3.2
scipy==1.11.4
six==1.16.0
sympy==1.12
tensorboard==2.15.1
tensorboard-data-server==0.7.2
threadpoolctl==3.2.0
tifffile==2023.9.26
toolz==0.12.0
torch==2.1.1
torchvision==0.16.1
tqdm==4.66.1
typing_extensions==4.8.0
tzdata==2023.3
umap-learn==0.5.5
urllib3==2.1.0
Werkzeug==3.0.1
zipp==3.17.0

File: tutorials/README.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+classes: /absolute/path/to/mnist_data/classes_mnist.csv
+affinity: /absolute/path/to/mnist_data/affinity_mnist.csv
+```
+
+In general, we recommend to work with absolute paths to avoid issues.
+
+To train AffinityVAE on our rotated MNIST dataset, run the following command:
+
+```bash
+python /absolute/path/to/affinity-vae/run.py --config_file /absolute/path/to/mnist_data/mnist_config.yml ---new_out
```
Comment (parkyed) [inline tutorials/README.md:70]: I think there is a typo in this line. ---new_out should be --new_out

File: tutorials/README.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+python -m venv env
+source env/bin/activate
+python -m pip install --upgrade pip
+python -m pip install -e .
+```
+
+### 2. Download the MNIST dataset
+
+The MNIST dataset can be downloaded from:
+https://figshare.com/articles/dataset/mnist_pkl_gz/13303457
```
Comment (marjanfamili) [inline tutorials/README.md:20]: This link comes out as page unavailable which confused me for a few minutes. is it easier to have this as a `wget command`

`wget https://figshare.com/ndownloader/files/25635053`

**Comment by marjanfamili** [comment]
Traceback (most recent call last):
  File "/Users/mfamili/work/tutorial_test/../affinity-vae/tutorials/mnist_saver.py", line 168, in <module>
    data = load_mnist(mnist_path)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mfamili/work/tutorial_test/../affinity-vae/tutorials/mnist_saver.py", line 16, in load_mnist
    f = gzip.open(path, 'rb')
        ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mfamili/anaconda3/lib/python3.11/gzip.py", line 58, in open
    binary_file = GzipFile(filename, gz_mode, compresslevel)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mfamili/anaconda3/lib/python3.11/gzip.py", line 174, in __init__
    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: 'mnist.pkl'
(new_tomlavae_env) (base) mfamili@585-D3FX0D9Y6Q tutorial_test % python ../affinity-vae/tutorials/mnist_saver.py --mnist_file mnist.pkl.gz --output_path . 
Traceback (most recent call last):
  File "/Users/mfamili/work/tutorial_test/../affinity-vae/tutorials/mnist_saver.py", line 168, in <module>
    data = load_mnist(mnist_path)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mfamili/work/tutorial_test/../affinity-vae/tutorials/mnist_saver.py", line 20, in load_mnist
    data = cPickle.load(f, encoding='bytes')
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mfamili/anaconda3/lib/python3.11/gzip.py", line 321, in peek
    return self._buffer.peek(n)
           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/mfamili/anaconda3/lib/python3.11/_compression.py", line 68, in readinto
    data = self.read(len(byte_view))
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mfamili/anaconda3/lib/python3.11/gzip.py", line 499, in read
    if not self._read_gzip_header():
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mfamili/anaconda3/lib/python3.11/gzip.py", line 468, in _read_gzip_header
    last_mtime = _read_gzip_header(self._fp)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mfamili/anaconda3/lib/python3.11/gzip.py", line 428, in _read_gzip_header
    raise BadGzipFile('Not a gzipped file (%r)' % magic)
gzip.BadGzipFile: Not a gzipped file (b'\x80\x02')

**Comment by marjanfamili** [comment]
Traceback (most recent call last):
  File "/Users/mfamili/work/tutorial_test/../affinity-vae/tutorials/mnist_saver.py", line 168, in <module>
    data = load_mnist(mnist_path)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mfamili/work/tutorial_test/../affinity-vae/tutorials/mnist_saver.py", line 20, in load_mnist
    data = cPickle.load(f, encoding='bytes')
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mfamili/anaconda3/lib/python3.11/gzip.py", line 321, in peek
    return self._buffer.peek(n)
           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/mfamili/anaconda3/lib/python3.11/_compression.py", line 68, in readinto
    data = self.read(len(byte_view))
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mfamili/anaconda3/lib/python3.11/gzip.py", line 499, in read
    if not self._read_gzip_header():
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mfamili/anaconda3/lib/python3.11/gzip.py", line 468, in _read_gzip_header
    last_mtime = _read_gzip_header(self._fp)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mfamili/anaconda3/lib/python3.11/gzip.py", line 428, in _read_gzip_header
    raise BadGzipFile('Not a gzipped file (%r)' % magic)
gzip.BadGzipFile: Not a gzipped file (b'\x80\x02')

**Comment by crangelsmith** [comment]
> Traceback (most recent call last): File "/Users/mfamili/work/tutorial_test/../affinity-vae/tutorials/mnist_saver.py", line 168, in data = load_mnist(mnist_path) ^^^^^^^^^^^^^^^^^^^^^^ File "/Users/mfamili/work/tutorial_test/../affinity-vae/tutorials/mnist_saver.py", line 20, in load_mnist data = cPickle.load(f, encoding='bytes') ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "/Users/mfamili/anaconda3/lib/python3.11/gzip.py", line 321, in peek return self._buffer.peek(n) ^^^^^^^^^^^^^^^^^^^^ File "/Users/mfamili/anaconda3/lib/python3.11/_compression.py", line 68, in readinto data = self.read(len(byte_view)) ^^^^^^^^^^^^^^^^^^^^^^^^^ File "/Users/mfamili/anaconda3/lib/python3.11/gzip.py", line 499, in read if not self._read_gzip_header(): ^^^^^^^^^^^^^^^^^^^^^^^^ File "/Users/mfamili/anaconda3/lib/python3.11/gzip.py", line 468, in _read_gzip_header last_mtime = _read_gzip_header(self._fp) ^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "/Users/mfamili/anaconda3/lib/python3.11/gzip.py", line 428, in _read_gzip_header raise BadGzipFile('Not a gzipped file (%r)' % magic) gzip.BadGzipFile: Not a gzipped file (b'\x80\x02')

This has been solved, Safari in automatically unzipped the file, this is not the case in Chrome. I have added a comment on the Readme on this.

**Comment by crangelsmith** [comment]
> Hi @crangelsmith , in working through the tutorial, in running the following line
> 
> python mnist_saver.py --mnist_file mnist.pkl.gz --output_path .
> 
> Traceback (most recent call last): File "/Users/ep/Documents/1_datasets/aff_vae/affinity-vae/tutorials/mnist_saver.py", line 179, in mnist_saver.run() File "/Users/ep/Documents/1_datasets/aff_vae/affinity-vae/tutorials/mnist_saver.py", line 128, in run image = rotate( File "/Users/ep/Documents/1_datasets/aff_vae/env/lib/python3.9/site-packages/scipy/ndimage/_interpolation.py", line 894, in rotate raise ValueError('input array should be at least 2D') ValueError: input array should be at least 2D
> 

Hi @parkyed, I haven't managed to reproduce this error. We realised last week if you download the file from Safari it gets automatically unzipped, so I wonder if that might happened to you. The error you get:

```
image = rotate( File "/Users/ep/Documents/1_datasets/aff_vae/env/lib/python3.9/site-packages/scipy/ndimage/_interpolation.py", line 894, in rotate raise ValueError('input array should be at least 2D') ValueError: input array should be at least 2D
``` 
makes me think that the data hasn't been loaded at all, and we are trying to rotate an empty array. Could this be the case?

**Review by parkyed — COMMENTED** [review]
Hi all. i've completed the review of the tutorial and have suggested edits to the markdown file. Let me know if you have questions.

**Review by parkyed — APPROVED** [review]
Resubmitting with approval.

**Pull request merged**